web/packages/teleport/src/Roles/Roles.tsx

```diff
@@ -43,17 +45,59 @@ export function RolesContainer() {
 }
 
 export function Roles(props: State) {
-  const { items, remove, save, attempt } = props;
-  const resources = useResources(items, templates);
+  const { remove, save, fetch } = props;
+  const [search, setSearch] = useState('');
+
+  const serverSidePagination = useServerSidePagination({
+    pageSize: 20,
+    fetchFunc: async (_, params) => {
+      const { items, startKey } = await fetch(params);
+      return { agents: items, startKey };
+    },
+    clusterId: '',
+    params: { search },
+  });
+  const { modifyFetchedData } = serverSidePagination;
+
+  const resources = useResources(
+    serverSidePagination.fetchedData.agents,
```

Comment:
can we put a note that we probably should rename `agents` to `resources` 😓

Reply:
Added, but I think we can go with even more generic `items` name. I saw that key being retuned from the backend.